### PR TITLE
examples/hwapi: Example showing best practices for HW API usage in apps.

### PR DESCRIPTION
**Pull request by pfalcon**
Showing and providing detailed instructions and motivation.

### CODE REVIEW

**Comment by pfalcon** [comment]
Recent discussion on HW API showed that people don't know what principles it's designed on and what's the intended way to use it. Of course, the maintainers are at fault by always keeping that in mind, referencing somewhere in tickets, but never coming out with a doc.

**Comment by peterhinch** [comment]
Very useful!

**Comment by rml4o** [comment]
+1

**Comment by profra** [comment]
+1 ... without watching all tickets and github one could be lost... up-to-date documentation and examples are preconditons for big family of users (not frustrated)... also some videos has been promised during KS campaign but Adafruit does good job in this area

File: examples/hwapi/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+which would work from a board to board, from a system to another systems.
+This is inherently a hard problem, because hardware is different from one
+board type to another, and even from examplar of board to another. For
+example, if your app requires an external LED, one user may connect it
+to one GPIO pin, while another user may find it much more convinient to
+use another pin. This of course applies to relays, buzzers, sensors, etc.
+
+With complications above in mind, it's still possible to write portable
+applications by using "low[est] denominator" subset of hardware API and
+following simple rules outlines below. The applications won't be able
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: outlined

File: examples/hwapi/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+to rely on advanced hardware capabilities of a particular boards and
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: board

File: examples/hwapi/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+will be limited to generic capabilities, but it's still possible to
+write many useful applications in such way, with the obvious benefit of
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: in such a way

File: examples/hwapi/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        utime.sleep_ms(500)
+
+
+To deploy this application to a particular board, a user will need:
+
+1. Edit `hwconfig.py` to adjust Pin and other hardware peripheral
+   parameters and locations.
+2. Actually deploy `hwconfig.py` and `app.py` to a board (e.g. copy to
+   board's filesystem, or build new firmware with these modules frozen
+   into).
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: into it

File: examples/hwapi/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+1. Look thru `hwconfig.py` to find a section which either exactly
+   matches your board, or the closest to it. Uncomment, and if any
+   adjustments required, apply them.
+
+It's important to keep in mind that adjustments may be always required,
+and that there may be users whose configuration doesn't match any of
+the available. So, always include a section or instructions for them.
+Consider for example that even on a supported board, user may want to
+blink not a on-board LED, but the one they connected externally.
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: an on-board

File: examples/hwapi/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+`hwconfig_*.py` - this assumes that a user would lazily deploy them all
+(or that application will be automatically installed, e.g. using MicroPython's
+`upip` package manager). The key point in this case remains the same as
+elaborated above - always assume there can, and will be a custom configuration,
+and it should be well supported. So, any automatic detection should be
+overridable by a user, and instructions how to do so are among the most
+important you may provide for your application.
+
+By following these best practices, you will use MicroPython at its full
+potential, and let users anjoy it too. Good luck!
```
Comment (dpgeorge) [inline examples/hwapi/README.md]: enjoy

**Review by dpgeorge — COMMENTED** [review]
Looks good!  I hope in the future we can add more stuff, eg i2c examples.

File: examples/hwapi/README.md (hunk 3 above)
Comment (pfalcon) [inline examples/hwapi/README.md]: Hmm, I even googled that up: https://www.usingenglish.com/forum/threads/2521-quot-in-such-way-quot-vs-quot-in-such-a-way-quot . Ok, "in such way" is colloquial.

**Comment by pfalcon** [comment]
Ok, thanks, typos fixed and patch merged. But, unfortunately, the example doesn't work as expected on ESP8266/ESP12 module. That's because it has an active-low LED, so instead of visually pleasing fade in-out, there's fading and blinking which kills all the effect. That's because of the way an LED works with PWM (or because of the way a human eye perceives it).

This leads us to https://github.com/micropython/micropython/issues/2603

**Pull request closed without merging**